File: PathSum2Backtracking.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    // intialize list to be returned
    List<List<Integer>> result = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        // call the recursive function
        helper(root, targetSum, 0, new ArrayList<>());
        return result;
    }
    private void helper(TreeNode root, int targetSum, int currentSum, List<Integer> path){
        // check the base condition
        if(root == null) return;
        // Add root's value to path list
        path.add(root.val);
        // update current sum
        currentSum += root.val;
        // check for leaf node and check if target is achieved
        if(root.left == null && root.right == null){
            if(currentSum == targetSum){
                result.add(new ArrayList<>(path));
            }
        }
        // call the funtion recursively for root's left child
        helper(root.left, targetSum, currentSum, path);
        // call the funtion recursively for root's right child
        helper(root.right, targetSum, currentSum, path);
        //backtracking
        path.remove(path.size() - 1);
    }
}

File: pathsum2_backtracking.py
#time complexity : O(n)
#Space Complexity : O(h) where h is height of tree
#did your code run on leetcode : yes

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    global result
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        global result
        result = []

        if root == None : return
        self.helper(root,targetSum,[],0)
        return result
        
        
    def helper(self, root: Optional[TreeNode], targetSum: int,path:List[int],currSum : int)->None:
            
            if root == None: return
            global result
            path.append(root.val)
            currSum += root.val
            
            #copy of path when currSum == targetSum is stored in the result due to which even if the all the nodes in path get deleted we will get the state at of path when targetSum is achieved
            if(root.left == None and root.right == None):
                if(currSum == targetSum):
                    result.append(path.copy())#copy of path stored in result

            self.helper(root.left,targetSum,path,currSum)
            self.helper(root.right,targetSum,path,currSum)
            del path[len(path)-1] #backtracking

